test/test_sparse.py

```diff
@@ -786,6 +786,54 @@ def test_shape(sparse_dims, nnz, with_size):
         test_shape(4, 3, [7, 7, 7, 3, 3, 3, 0])
         test_shape(4, 0, [0, 0, 7, 3, 3, 3, 0])
 
+    @coalescedonoff
+    @dtypes(torch.double, torch.cdouble)
+    def test_permute(self, device, dtype, coalesced):
+        # trivial checks
+        s = torch.rand(3, 3, 3, device=device, dtype=dtype).to_sparse()
+        with self.assertRaisesRegex(RuntimeError, "does not match the length"):
+            s.permute(dims=(1, 0))
+        with self.assertRaisesRegex(RuntimeError, "duplicate dims"):
+            s.permute(dims=(1, 1, 1))
+
+        def test_shape(sparse_dims, nnz, with_size):
+            ndim = len(with_size)
+            valid_sparse_dims = torch.arange(-ndim, -ndim + sparse_dims)
+            valid_dense_dims = torch.arange(-ndim + sparse_dims, 0)
+
+            for dims in itertools.permutations(range(-ndim, 0)):
+                s = self._gen_sparse(sparse_dims, nnz, with_size, dtype, device, coalesced)[0]
+                d = self.safeToDense(s)
+
+                dims_sparse, _ = torch.tensor(dims[:sparse_dims]).sort()
+                dims_dense, _ = torch.tensor(dims[sparse_dims:]).sort()
+
+                if (valid_sparse_dims == dims_sparse).all() and (valid_dense_dims == dims_dense).all():
+                    # if valid permutation, test for correctness
+                    s_permuted = s.permute(dims)
+                    self.assertEqual(s_permuted, d.permute(dims))
+
+                    # if s is coalesced, and perm does not touch 0-dim,
+                    # the result has to be coalesced as well
+                    if dims[0] == 0:
+                        self.assertEqual(s_permuted.is_coalesced(), s.is_coalesced())
+                    else:
+                        self.assertFalse(s_permuted.is_coalesced())
+
+                    gradcheck(lambda t: t.permute(dims).to_dense(), s.requires_grad_(True), check_sparse_nnz=True)
+                else:
+                    # otherwise check if exception is thrown
+                    fail_message = "transpositions between sparse and dense dimensions are not allowed"
+                    with self.assertRaisesRegex(RuntimeError, fail_message):
+                        s.permute(dims)
+
+        test_shape(2, 3, [2, 3, 4, 5])
+        test_shape(2, 3, [2, 2, 0])
+        # For some reason the case below fails for gradcheck:
+        # grads have indices of shape (3, 0), not (2, 0) as it should be.
+        # TODO: file an issue
+        # test_shape(2, 0, [0, 0, 2])
```

Comment:
@albanD , @soulitzer , any quick ideas?

Reply:
The [derivative for permute](https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L1235-L1236) is in turn again [permute again](https://github.com/pytorch/pytorch/blob/16f30b494c43d9c4742caf62a18d282a59e0f2e7/torch/csrc/autograd/FunctionsManual.cpp#L516-L524). Maybe the most likely explanation is a bug in the implementation of permute above. You should be able to trace the calls to permute_sparse_coo for the particular inputs and see if you can reproduce. Another way to debug this would be to add a runtime check at the end of permute_sparse_coo and use to/from Strided conversion to check the results.

Reply:
I am not sure, the forward is correct, it matches the dense implementation. I have no idea where index of shape with size 3 even comes from, the original input's shape does not have 3 at all...

Reply:
Hmm, or it could be an `IntArrayRef` issue with memory overlap between destroyed objects. Pearu did hit it at some point, will have a look.

Reply:
@cpuhrsch , the issue is that returned `grad` is sparse, but it has wrong shape, that is what gradcheck tells.

Reply:
@cpuhrsch , there must be something wrong the graph, I guess. Check out the following. If I modify `permute_backward` to be:
```c++
Tensor permute_backwards(const Tensor& grad, IntArrayRef fwd_dims, const Tensor& self) {
  // invert the permutation
  auto ndims = fwd_dims.size();
  std::vector<int64_t> dims(ndims);
  for (const auto i : c10::irange(ndims)) {
    dims[at::maybe_wrap_dim(fwd_dims[i], ndims)] = i;
  }
  auto res = grad.permute(dims);
  std::cout << "grad: " << grad << std::endl;
  std::cout << "grad.permute: " << res << std::endl;
  std::cout << "dims: " << fwd_dims << std::endl;
  std::cout << "self : " << self << std::endl;
  std::cout << "self.permute : " << self.permute(fwd_dims) << std::endl;
  std::cout << std::endl;
  return res;
}

```
and then run the failing case, I get:
```
grad: [ SparseCUDADoubleType{}                                                                                                                                               
indices:                                                                                                                                                                     
[ CUDALongType{3,0} ]                                                                 
values:                                                                                                                                                                      
[ CUDADoubleType{0} ]                                                                                                                                                        
size:                                                                                                                                                                        
[0, 0, 2]                                                                                                                                                                    
]                                                                                                                                                                            
grad.permute: [ SparseCUDADoubleType{}                                                                                                                                       
indices:                                                                                                                                                                     
[ CUDALongType{3,0} ]                                                                                                                                                        
values:                                                                                                                                                                      
[ CUDADoubleType{0} ]                                                                                                                                                        
size:                                                                                                                                                                        
[0, 0, 2]                                                                             
]                                                                                                                                                                            
dims: [-3, -2, -1]                                                                                                                                                           
self : [ SparseCUDADoubleType{}                                                       
indices:                                                                                                                                                                     
[ CUDALongType{2,0} ]                                                                                                                                                        
values:                                                                               
[ CUDADoubleType{0,2} ]                                                                                                                                                      
size:                                                                                                                                                                        
[0, 0, 2]                                                                                                                                                                    
]                                                                                                                                                                            
self.permute : [ SparseCUDADoubleType{}                                                                                                                                      
indices:                                                                                                                                                                     
[ CUDALongType{2,0} ]                               
values:                                             
[ CUDADoubleType{0,2} ]                             
size:                                               
[0, 0, 2]                                           
]                                      
```
Note that the in-flowing grad shape is wrong.